### PR TITLE
Add [push|pop]_[front|back] for Array

### PR DESCRIPTION
**Pull request by bernhardmgruber**
Add `[push|pop]_[front|back]` for `Array`.

### CODE REVIEW

File: include/llama/Array.hpp [hunk 1]
```diff
@@ -89,6 +89,56 @@ namespace llama
 
     template <typename First, typename... Args>
     Array(First, Args... args) -> Array<First, sizeof...(Args) + 1>;
+
+    template <typename T, std::size_t N>
+    LLAMA_FN_HOST_ACC_INLINE constexpr auto push_front(Array<T, N> a, T v) -> Array<T, N + 1>
```
Comment (j-stephan) [inline include/llama/Array.hpp:94]: Is passing by value intended here?

File: include/llama/Array.hpp [hunk 2]
```diff
@@ -89,6 +89,56 @@ namespace llama
...
+    LLAMA_FN_HOST_ACC_INLINE constexpr auto pop_back(Array<T, N> a) -> Array<T, N - 1>
+    {
+        Array<T, N - 1> r{};
+        for (auto i = 0; i < N - 1; i++)
+            r[i] = a[i];
+        return r;
+    }
+
+    template <typename T>
+    LLAMA_FN_HOST_ACC_INLINE constexpr auto pop_back(Array<T, 0> a) -> Array<T, 0>
```
Comment (j-stephan) [inline include/llama/Array.hpp]: What are the intended semantics here? To me this sounds like a case for `static_assert`.

**Review by j-stephan — APPROVED** [review]
Looking good, a few questions:

File: include/llama/Array.hpp (hunk 1 above)
Comment (bernhardmgruber) [reply include/llama/Array.hpp:94]: Yes. While generally passing large arrays per value might cause expensive copies, the arrays in LLAMA are quite small usually. Most oftenly, these arrays hold the index space dimensions so usually just have 1 or maybe 2 or 3 elements. Furthermore, the call is forced inline as well, so a function call should never occur.

File: include/llama/Array.hpp (hunk 2 above)
Comment (bernhardmgruber) [reply include/llama/Array.hpp]: I pulled this out of a different PR, where I need this precise semantic. But we could argue it should be a differently named function. I will probably change the name then.

Comment (bernhardmgruber) [reply include/llama/Array.hpp]: I found a way to not need this special case. I removed it.

File: include/llama/Array.hpp (hunk 1 above)
Comment (j-stephan) [reply include/llama/Array.hpp:94]: Okay. Another question: What is the intended use case? In the test you just pass in some rvalues. Is this always the case in the real world? If not, is it intended to have two arrays after the operation?

Comment (bernhardmgruber) [reply include/llama/Array.hpp:94]: The way it's currently used is to add a `size_t` to an array of `size_t`s. The old and the new array continue living.

**Pull request merged**